extensions/barrage/src/main/java/io/deephaven/extensions/barrage/util/PythonTableDataService.java

```diff
@@ -863,4 +929,35 @@ public long size() {
     private static class Subscription {
         SafeCloseable cancellationCallback;
     }
+
+    // this tool is used to simplify backend asynchronous RPC patterns for otherwise synchronous operations
+    private static class AsyncState<T> {
+        private T result;
+        private RuntimeException error;
+
+        public synchronized void setResult(@NotNull final T result) {
+            this.result = result;
+            notifyAll();
+        }
+
+        public synchronized void setError(@NotNull final RuntimeException error) {
+            this.error = error;
+            notifyAll();
+        }
+
+        public synchronized T awaitResult(@NotNull final Function<Exception, RuntimeException> errorMapper) {
```

Comment:
Seems weird that `errorMapper` accepts an `Exception`, but `error` is a `RuntimeException`. Maybe we want to be more permissive in specifying `error`, or just get rid of `errorMapper`?

Reply:
I wanted the mapper to guarantee that we can throw it without declaring that the method call throws. I wanted to be permissive on the way into the mapper so we could setError to an `IOException`, for example.

However, I do see that I ended up wrapping the `IOException` in an `UncheckedDeephavenException` anyway .. and then that might be wrapped in a `TableDataException`.

The benefit of the mapper is to capture that the error may have originated off-thread and therefore both stack traces may have real value to the user. As well as, to move specifics about the call into the mapper (for consistency in error message) and allow for the internal error to not necessarily care the origin of the request.

Reply:
Ah, actually, the `Exception` entry to the helper is because of the interrupted exception in awaitResult. Will make the holder more permissible.